### PR TITLE
WIP : multi panels (for question comment purpose)

### PR DESCRIPTION
**Pull request by kosamari**
Just making PR because commenting UI is better on PR.

Approach I took was to accept any HTML elements as children of `<multi-panels>` 

`<multi-panels>` will simply take even indexed elements as heading, and element following(thus odd index) as content.  Allowing user of this component more freedom in markup.

note: I don't really understand TypeScript(yet) and this might very well be not great code(yet) !

### CODE REVIEW

File: components/MultiPanel/index.ts [hunk 1]
```diff
@@ -1,12 +1,183 @@
...
+
+  // remove EventListeners
+  disconnectedCallback() {
+    this.removeEventListener('click', this._onClick);
+    this.removeEventListener('keydown', this._onKeyDown);
   }
+  
+  // Click event handler
+  _onClick(event) {
+    const heading : Element = event.target;
```
Comment (kosamari) [inline components/MultiPanel/index.ts]: I don't know what to do with types when we are doing simple event listeners...
When I do 
```
_onClick(event : Event) {
  const heading : Element = event.target
}
```
then type script yells `Type 'EventTarget | null' is not assignable to type 'Element'`

File: components/MultiPanel/index.ts [hunk 2]
```diff
@@ -1,12 +1,183 @@
...
+      // Any other key press is ignored and passed back to the browser.
+      default:
+        return;
+    }
+
+    event.preventDefault();
+    currentHeading.setAttribute('tabindex', '-1');
+    if (newHeading) {
+      newHeading.setAttribute('tabindex', '0');
+      newHeading.focus();
```
Comment (kosamari) [inline components/MultiPanel/index.ts]: Type script is yelling `Property 'focus' does not exist on type 'Element'` here... but I can't find the way to make return of `_allHeadings` to `HTMLElement` ...

File: components/MultiPanel/index.ts [hunk 3]
```diff
@@ -1,12 +1,183 @@
...
+      case 'Home':
+        newHeading = this._firstHeading();
+        break;
+
+      case 'End':
+        newHeading = this._lastHeading();
+        break;
+
+      case 'Enter':
+      case ' ':
```
Comment (kosamari) [inline components/MultiPanel/index.ts]: If we want to support IEs and FF before 37, then `Spacebar` might be needed here.
https://developer.mozilla.org/en-US/docs/Web/API/KeyboardEvent/key/Key_Values

Comment (jakearchibald) [reply components/MultiPanel/index.ts]: We don't need to support those browsers for this project, but in terms of using this component outside this project, I'll leave it up to you.

File: components/MultiPanel/index.ts [hunk 4]
```diff
@@ -1,12 +1,183 @@
...
+
+    // Watch for children changes.
+    new MutationObserver(() => this._childrenChange())
+      .observe(this, { childList: true });
   }
 
   connectedCallback () {
+    this._childrenChange();
+
+    const children : Element[] = Array.from(this.children);
```
Comment (jakearchibald) [inline components/MultiPanel/index.ts]: The following code assumes the internals of the multi-panel are complete once the element is connected, and never change after that. Ideally that shouldn't be assumed. It'd be better to observe this stuff in `_childrenChange` and react to elements coming and going.

File: components/MultiPanel/index.ts [hunk 5]
```diff
@@ -1,12 +1,183 @@
...
+    for (let i : number = 0; i < children.length; i++){
+      const child : Element = children[i]
+
+      if(i % 2 === 0){  
+        // even index = heading element
+        child.classList.add('panel-heading');
+        
+        // for A11y
+        child.id = `panel-heading-${i}`;
```
Comment (jakearchibald) [inline components/MultiPanel/index.ts]: If the child already has an ID we shouldn't overwrite it.

File: components/MultiPanel/index.ts [hunk 6]
```diff
@@ -1,12 +1,183 @@
...
+    // Watch for children changes.
+    new MutationObserver(() => this._childrenChange())
+      .observe(this, { childList: true });
   }
 
   connectedCallback () {
+    this._childrenChange();
+
+    const children : Element[] = Array.from(this.children);
+    for (let i : number = 0; i < children.length; i++){
```
Comment (mathiasbynens) [inline components/MultiPanel/index.ts]: Nit: use `forEach` instead of a hand-written loop. `forEach` is highly optimized for proper arrays in V8, so it’s likely faster, and it still gives you access to the `index` variable (unlike `for`-`of`).

File: components/MultiPanel/index.ts [hunk 7]
```diff
@@ -1,12 +1,183 @@
...
+    if (!this._isHeading(currentHeading)) {
+      return;
+    }
+
+    // don’t handle modifier shortcuts used by assistive technology.    
+    if (event.altKey) {
+      return;
+    }
+    let newHeading;
+    switch (event.key) {
```
[the remark this answers is not in the document]
Comment (mathiasbynens) [reply components/MultiPanel/index.ts]: Nice work — I love it! So readable 👍🏻

File: components/MultiPanel/index.ts [hunk 8]
```diff
@@ -1,12 +1,183 @@
...
+        child.classList.add('panel-content');
+
+        // for A11y
+        child.id = `panel-content-${i-1}`;
+        child.setAttribute('aria-labelledby', `panel-heading-${i-1}`);
+      } 
+    }
+
+    // make the first heading focusable.
+    children[0].setAttribute('tabindex', '0');
```
Comment (jakearchibald) [inline components/MultiPanel/index.ts]: Is https://developer.mozilla.org/en-US/docs/Web/HTML/Element/details a better primitive here in terms of accessibility? @surma did you ever look into this?

It might be that `<details>` buggers up too much when it comes to animation.

File: components/MultiPanel/index.ts [hunk 9]
```diff
@@ -1,12 +1,183 @@
...
+  // return list of panel heading elements
+  _allHeadings () {
+    return Array.from(this.querySelectorAll('.panel-heading'));
+  }
+
+  // returns headding that is before currently selected one.
+  _prevHeading () {
+    const headings : Element[] = this._allHeadings();
+    let newIdx : number = headings.findIndex(headings => headings === document.activeElement) - 1;
+    return headings[(newIdx + headings.length) % headings.length];
```
Comment (mathiasbynens) [inline components/MultiPanel/index.ts]: Maybe document the edge cases here? E.g. when headings contains a single element, this code just returns that.

File: components/MultiPanel/index.ts (hunk 6 above)
Comment (mathiasbynens) [reply components/MultiPanel/index.ts]: (https://v8project.blogspot.com/2017/09/elements-kinds-in-v8.html if you’re interested in the background)

Comment (jakearchibald) [reply components/MultiPanel/index.ts]: Hah, I recommended for-of below. Which is better? Fwiw I think we should be optimising for readability over what Chrome does fastest on a particular day.

File: components/MultiPanel/index.ts [hunk 10]
```diff
@@ -1,12 +1,183 @@
...
+    this.removeEventListener('click', this._onClick);
+    this.removeEventListener('keydown', this._onKeyDown);
   }
+  
+  // Click event handler
+  _onClick(event) {
+    const heading : Element = event.target;
+    if (this._isHeading(heading)) {
+      this._expand(heading);
+    }
```
Comment (jakearchibald) [inline components/MultiPanel/index.ts]: This breaks if the user clicks an element inside the heading. As in:

```html
<h1>Hello <em>world</em></h1>
```

Clicking `world` won't work.

Instead of `_isHeading`, maybe write something like `_getClosestHeading`? `el.closest('multi-panel > *')` will help here.

`_getClosestHeading` should return undefined/null if the user didn't click within a heading.

File: components/MultiPanel/index.ts (hunk 6 above)
Comment (mathiasbynens) [reply components/MultiPanel/index.ts]: > Fwiw I think we should be optimising for readability over what Chrome does fastest on a particular day.

To be clear, I fully agree with this, and this matches our general V8 messaging.

In this case, a handwritten `for` loop expresses _implementation_, whereas `for-of` + `entries` or `forEach` would express _intent_. Expressing intent often leads to more readable code, IMHO.

`forEach` likely being faster than `for` in this case is an indirect result of that (and a nice little bonus). By expressing intent rather than implementation, you generally give the JS engine more information to work with, often leading to greater behind-the-scenes optimization potential.

File: components/MultiPanel/index.ts [hunk 11]
```diff
@@ -1,12 +1,183 @@
...
+    new MutationObserver(() => this._childrenChange())
+      .observe(this, { childList: true });
   }
 
   connectedCallback () {
+    this._childrenChange();
+
+    const children : Element[] = Array.from(this.children);
+    for (let i : number = 0; i < children.length; i++){
+      const child : Element = children[i]
```
[the remark this answers is not in the document]
Comment (mathiasbynens) [reply components/MultiPanel/index.ts]: Between that or `forEach((child, index) => { … })`, I don’t really have a preference. I do think either of those are nicer than the hand-written `for` loop though

File: components/MultiPanel/index.ts (hunk 1 above)
Comment (jakearchibald) [reply components/MultiPanel/index.ts]: You can go one of two ways with this. If you think `event.target` can only ever be an `Element`, you can do:

```ts
const heading = event.target as Element;
```

See https://www.typescriptlang.org/docs/handbook/basic-types.html#type-assertions.

However, if TypeScript is being helpful, you can make use of [type guards](https://www.typescriptlang.org/docs/handbook/advanced-types.html#type-guards-and-differentiating-types).

```ts
if (!(event.target instanceof Element)) return;
const heading: Element = event.target;
```

The above works, because TypeScript infers that `event.target` must be an `Element`.

File: components/MultiPanel/index.ts [hunk 12]
```diff
@@ -1,12 +1,183 @@
...
+  }
+
+  // KeyDown event handler
+  _onKeyDown(event) {
+    const currentHeading : Element = event.target;
+    
+    // if clicke event is not on heading element, ignore
+    if (!this._isHeading(currentHeading)) {
+      return;
+    }
```
Comment (jakearchibald) [inline components/MultiPanel/index.ts]: Might be worth looking at [user defined type guards](https://www.typescriptlang.org/docs/handbook/advanced-types.html#user-defined-type-guards) here.

That way, you can tell TypeScript that, if `_isHeading` returns true, the type of `currentHeading` must be `HTMLElement` or whatever.

File: components/MultiPanel/index.ts [hunk 13]
```diff
@@ -1,12 +1,183 @@
...
+
+    // don’t handle modifier shortcuts used by assistive technology.    
+    if (event.altKey) {
+      return;
+    }
```
Comment (jakearchibald) [inline components/MultiPanel/index.ts]: Good catch!

File: components/MultiPanel/index.ts (hunk 12 above)
Comment (jakearchibald) [reply components/MultiPanel/index.ts]: Note: if you do the `_closestHeading` thing I mentioned above, `_isHeading` is still the best thing to use in this case.

File: components/MultiPanel/index.ts [hunk 14]
```diff
@@ -1,12 +1,183 @@
...
+  _nextHeading () {
+    const headings : Element[] = this._allHeadings();
+    let newIdx  : number = headings.findIndex(heading => heading === document.activeElement) + 1;
+    return headings[newIdx % headings.length];
+  }
+
+  // returns first heading in multi-panel.
+  _firstHeading () {
+    const headings : Element[] = this._allHeadings();
+    return headings[0];
```
Comment (jakearchibald) [inline components/MultiPanel/index.ts]: I think this would be faster:

```ts
const children = this.children;
if (children.length % 1) {
  // Odd number of children. Last one must be a heading.
  return children[children.length - 1];
}
// Otherwise, the second-last child is the heading.
return children[children.length - 2];
```

File: components/MultiPanel/index.ts [hunk 15]
```diff
@@ -1,12 +1,183 @@
...
+  } 
+
+  // returns last heading in multi-panel.
+  _lastHeading () {
+    const headings : Element[] = this._allHeadings();
+    return headings[headings.length - 1];
```
Comment (jakearchibald) [inline components/MultiPanel/index.ts]: I think this would be faster:

```ts
const children = this.children;
if (children.length % 1) {
  // Odd number of children. Last one must be a heading.
  return children[children.length - 1];
}
// Otherwise, the second-last child is the heading.
return children[children.length - 2];
```

File: components/MultiPanel/index.ts (hunk 5 above)
Comment (jakearchibald) [reply components/MultiPanel/index.ts]: Ohh, and you're going to get duplication of IDs if you have two `<multi-panel>`s. Maybe generate a random ID?

File: components/MultiPanel/index.ts (hunk 11 above)
[the remark this answers is not in the document]
Comment (jakearchibald) [reply components/MultiPanel/index.ts]: No preference here either.

File: components/MultiPanel/index.ts (hunk 6 above)
Comment (developit) [reply components/MultiPanel/index.ts]: While we're on the subject: generally when iterating over children, it's fastest to use traversal:

```js
let child = this.firstElementChild;
while (child) {
  // stuff
  child = child.nextElementSibling;
}
```

I find this offers a nice balance of readability and performance, since it avoids the array casting entirely.

File: components/MultiPanel/index.ts (hunk 11 above)
[the remark this answers is not in the document]
Comment (developit) [reply components/MultiPanel/index.ts]: Why use an Array at all?

```js
let child = this.firstElementChild;
while (child) {
  // stuff
  child = child.nextElementSibling;
}
```

- nice balance of readability and performance
- avoids the array casting entirely

[the remark this answers is not in the document]
Comment (mathiasbynens) [reply components/MultiPanel/index.ts]: Array operations are heavily optimized in V8 for actual arrays. The one-off cost of casting to an array is usually worth the later performance benefits when operating on the array (as opposed to an array-like object). (shameless https://www.youtube.com/watch?v=m9cTaYI95Zc plug in case anyone wants more info)

TL;DR Don’t think of the `Array.from()` as overhead.

**Comment by kosamari** [comment]
Status : 
Current code has all refraction of the previous code review (minus `<details>` consideration).

This component currently assumes there is always a set of `heading` and  `content`.
and expands corresponding content via `nextElementSibling`.

If developer using this component does something like below (per @jakearchibald's review) and user click on heading before content is created, it should be fine (simply no`nextElementSibling`)
```
multiPanel.append(parse(`<h1>Hello!</h1>`));
const data = await fetch(content).then(r => r.json());
multiPanel.append(parse(data.html));
```
 
however, if developer inserts element in async way like above in the middle of multi panel children... it will mess up which panel to be 'expanded'.

Not sure expecting dev user to "always append 2 elements together" is a reasonable expectation (code will be simpler !) or we should bullet proof this with unique ID approach and control element gets expanded based on ID.

note: currently unique id is assigned just for A11y but not for control.

File: components/MultiPanel/index.ts [hunk 16]
```diff
@@ -1,12 +1,197 @@
...
+  _childrenChange () {
+    if (this.children.length % 2 !== 0) {
+      console.error('detected odd number of elements inside multi-panel, please make sure you have heading/content pair')
+    }
+
+    const children : Element[] = Array.from(this.children);
+    let tabindexAlreadySet : boolean = false;
+    let randomId : string;
+
+    children.forEach((child, i) => {
```
Comment (jakearchibald) [inline components/MultiPanel/index.ts]: I think this would be more readable as:

```js
for (let i = 0; i < children.length; i += 2) {
  const heading = children[i];
  const content = children[i+1];
  const randomId = Math.random().toString(36).substr(2, 9);

  // …
}
```

As you don't need to do the odd/even check, and `randomId` gets better scoping.

File: components/MultiPanel/index.ts [hunk 17]
```diff
@@ -1,12 +1,197 @@
...
+      if(i % 2 === 0){
+        // even index = heading element
+
+        // if panel-heading class is already assigned to the element (= new child was added later),
```
Comment (jakearchibald) [inline components/MultiPanel/index.ts]: Earlier?

File: components/MultiPanel/index.ts [hunk 18]
```diff
@@ -1,12 +1,197 @@
...
+        // skip the ID/tabindex setting activities & just flag tabindexAlreadySet
+        if (!child.classList.contains('panel-heading')){
```
Comment (jakearchibald) [inline components/MultiPanel/index.ts]: If there's an if statement with a short part & a long part, I think it's more readable to handle the short part first. In some cases it means you can avoid the `else`. In this case:

```js
if (child.classList.contains('panel-heading')) {
  tabIndexAlreadySet = true;
  return;
}

child.classList.add('panel-heading');
// …
```

The `return` will need to be `continue` if you change this to a for-loop.

File: components/MultiPanel/index.ts [hunk 19]
```diff
@@ -1,12 +1,197 @@
...
+    const children : Element[] = Array.from(this.children);
+    let tabindexAlreadySet : boolean = false;
+    let randomId : string;
+
+    children.forEach((child, i) => {
+      if(i % 2 === 0){
+        // even index = heading element
+
+        // if panel-heading class is already assigned to the element (= new child was added later),
+        // skip the ID/tabindex setting activities & just flag tabindexAlreadySet
```
Comment (jakearchibald) [inline components/MultiPanel/index.ts]: I think this strategy has problems if you replace just a heading, or just the content. Or, if you swap some headings around but not the content.

File: components/MultiPanel/index.ts (hunk 3 above)
Comment (kosamari) [reply components/MultiPanel/index.ts]: handled by adding `case 'Spacebar':`

File: components/MultiPanel/index.ts (hunk 16 above)
Comment (kosamari) [reply components/MultiPanel/index.ts]: handled this in suggested way.

**Comment by kosamari** [comment]
closing this in lieu of #95
